src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -6116,12 +6126,12 @@ private void MoveDirectoryInfoUnchecked(DirectoryInfo directory, string destinat
             {
                 if (InternalTestHooks.ThrowExdevErrorOnMoveDirectory)
                 {
-                    throw new IOException("Invalid cross-device link");
+                    throw new IOException("Invalid cross-device link", hresult: MOVE_FAILED_ERRORS[0]);
                 }
 
                 directory.MoveTo(destinationPath);
             }
-            catch (IOException)
+            catch (IOException e) when (MOVE_FAILED_ERRORS.Contains(e.HResult))
```

Comment:
We could simplify the code and exclude allocations:
```suggestion
#if UNIX
            // This is the errno returned by the rename() syscall
            // when an item is attempted to be renamed across filesystem mount boundaries.
            catch (IOException e) when (e.HResult == 18)
#else
            // 0x80070005 ACCESS_DENIED is returned when trying to move files across volumes like DFS
            // 0x80131620 is returned if the source and destination do not have the same root path
            catch (IOException e) when (e.HResult == -2147024891)
            catch (IOException e) when (e.HResult == -2146232800)
#endif
```

Reply:
Good suggestion, will update